__tests__/host.spec.ts

```diff
@@ -0,0 +1,111 @@
+import { afterAll, beforeAll, test, expect, jest } from "@jest/globals";
+import * as ts from "typescript";
+import * as path from "path";
+import { normalizePath as normalize } from "@rollup/pluginutils";
+import { remove, ensureDir, writeFile } from "fs-extra";
+
+import { setTypescriptModule } from "../src/tsproxy";
+import { LanguageServiceHost } from "../src/host";
+
+setTypescriptModule(ts);
+
+const defaultConfig = { fileNames: [], errors: [], options: {} };
+
+const unaryFunc = "const unary = (x: string): string => x.reverse()";
+const unaryFuncSnap = { text: unaryFunc };
+
+// host.ts uses `/` normalized path
+const local = (x: string) => normalize(path.resolve(__dirname, x));
+const testDir = local("__temp/host");
+const testFile = `${testDir}/file.ts`;
+const nonExistent = `${testDir}/this-does-not-exist.ts`;
+
+afterAll(() => remove(testDir));
+beforeAll(async () => {
+	await ensureDir(testDir);
+	await writeFile(testFile, unaryFunc, "utf8");
+});
+
+test("LanguageServiceHost", async () => {
+	const testOpts = { test: "this is a test" }
+	const config = { ...defaultConfig, options: testOpts };
+	const transformers = [() => ({})];
+	const host = new LanguageServiceHost(config, transformers, testDir);
+
+	// test core snapshot functionality
+	expect(host.getScriptSnapshot(testFile)).toEqual(unaryFuncSnap);
+	expect(host.getScriptVersion(testFile)).toEqual("1");
+
+	expect(host.setSnapshot(testFile, unaryFunc)).toEqual(unaryFuncSnap); // version 2
+	expect(host.getScriptSnapshot(testFile)).toEqual(unaryFuncSnap); // get from dict
+	expect(host.getScriptVersion(testFile)).toEqual("2");
+
+	expect(host.getScriptSnapshot(nonExistent)).toBeFalsy();
+	expect(host.getScriptVersion(nonExistent)).toEqual("0");
+
+	expect(host.getScriptFileNames()).toEqual([testFile]);
+
+	host.reset(); // back to version 1
+	expect(host.setSnapshot(testFile, unaryFunc)).toEqual(unaryFuncSnap);
+	expect(host.getScriptVersion(testFile)).toEqual("1");
+
+	// test fs functionality, which just uses tsModule.sys equivalents
+	expect(host.getCurrentDirectory()).toEqual(testDir);
+	expect(host.getDirectories(testDir)).toEqual([]);
+	expect(host.directoryExists(nonExistent)).toBeFalsy();
+	expect(host.fileExists(nonExistent)).toBeFalsy();
+	expect(host.fileExists(testFile)).toBeTruthy();
+	expect(host.readDirectory(testDir)).toEqual([testFile]);
+	expect(host.readFile(nonExistent)).toBeFalsy();
+	expect(host.readFile(testFile)).toEqual(unaryFunc);
+	expect(host.useCaseSensitiveFileNames()).toBe(process.platform === "linux");
+
+	// test misc functionality
+	expect(host.getCompilationSettings()).toEqual(testOpts);
+	expect(host.getDefaultLibFileName({})).toEqual(
+		require.resolve("typescript/lib/lib.d.ts")
+	);
```

Comment:
Ok I tried figuring this out for a few hours (while working on a bunch of other stuff too -- some more refactors here and more PRs coming!) and it seems that `tsModule.getDefaultLibFilePath({})` just returns a wacky half-normalized path on Windows???

```
    Expected: "D:\\a\\rollup-plugin-typescript2\\rollup-plugin-typescript2\\node_modules\\typescript\\lib\\lib.d.ts"
    Received: "D:\\a\\rollup-plugin-typescript2\\rollup-plugin-typescript2\\node_modules\\typescript\\lib/lib.d.ts"
 ```
 
 Screenshot from the Actions logs (which ran a bunch of times on this...):
 
![Screen Shot 2022-05-10 at 1 08 29 AM](https://user-images.githubusercontent.com/4970083/167547223-c36cca06-2ff1-4fa5-8083-6eac554f6ed7.png)

Reply:
Oh... Looked at the TS source code and this indeed appears to be a bug [in TS](https://github.com/microsoft/TypeScript/blob/8f56f6b49d2fb282ff95df64f62dab2947004e95/src/services/services.ts#L2881)... was not expecting that 😅  
Thought either I was doing something wrong or the `host.ts` file had a bug in for the longest time.... the code is like 7 years old too, very surprised this hasn't come up before?? (and no existing issues on it either)

Gonna submit a PR to TS tomorrow and put a workaround in here then too

Reply:
With the amount of path normalization bugs out there every lib should basically assume everything is possible :)

Reply:
You'd think it would've been discovered in 7 years though in such a popular tool like TS itself! Guess not even TS is safe 😅 

Filed issue https://github.com/microsoft/TypeScript/issues/49050 and PR https://github.com/microsoft/TypeScript/pull/49051, will write up a tiny workaround for this till that gets merged, released, etc in the meantime

Reply:
Just checked `endsWith("lib.d.ts")` as a workaround for now -- left a TODO comment referencing this issue for future reference for a better test.